File: regions/system/library/go/migration/declarative.go
package migration

import (
	"fmt"
	"strings"

	"github.com/BurntSushi/toml"
)

type tableDef struct {
	Table tableSpec `toml:"table"`
}

type tableSpec struct {
	Name    string       `toml:"name"`
	Columns []columnSpec `toml:"columns"`
}

type columnSpec struct {
	Name       string  `toml:"name"`
	DataType   string  `toml:"type"`
	PrimaryKey bool    `toml:"primary_key"`
	Nullable   *bool   `toml:"nullable"` // pointer to distinguish unset (default true)
	Default    *string `toml:"default"`
	Unique     bool    `toml:"unique"`
	References *string `toml:"references"`
}

func (c columnSpec) isNullable() bool {
	if c.Nullable == nil {
		return true
	}
	return *c.Nullable
}

// TomlToCreateSQL converts a TOML table definition to CREATE TABLE SQL.
func TomlToCreateSQL(tomlStr string) (string, error) {
	var def tableDef
	if err := toml.Unmarshal([]byte(tomlStr), &def); err != nil {
		return "", fmt.Errorf("TOML parse error: %w", err)
	}

	var colDefs []string
	var primaryKeys []string

	for _, col := range def.Table.Columns {
		parts := []string{fmt.Sprintf("%s %s", col.Name, col.DataType)}

		if !col.isNullable() && !col.PrimaryKey {
			parts = append(parts, "NOT NULL")
		}

		if col.Unique {
			parts = append(parts, "UNIQUE")
		}

		if col.Default != nil {
			parts = append(parts, fmt.Sprintf("DEFAULT %s", *col.Default))
		}

		if col.References != nil {
			parts = append(parts, fmt.Sprintf("REFERENCES %s", *col.References))
		}

		if col.PrimaryKey {
			primaryKeys = append(primaryKeys, col.Name)
		}

		colDefs = append(colDefs, strings.Join(parts, " "))
	}

	if len(primaryKeys) > 0 {
		colDefs = append(colDefs, fmt.Sprintf("PRIMARY KEY (%s)", strings.Join(primaryKeys, ", ")))
	}

	return fmt.Sprintf("CREATE TABLE %s (\n  %s\n);", def.Table.Name, strings.Join(colDefs, ",\n  ")), nil
}

File: regions/system/library/typescript/migration/src/declarative.ts
import { parse } from 'smol-toml';
import { MigrationError } from './index.js';

interface ColumnSpec {
  name: string;
  type: string;
  primary_key?: boolean;
  nullable?: boolean;
  default?: string;
  unique?: boolean;
  references?: string;
}

interface TableSpec {
  name: string;
  columns: ColumnSpec[];
}

interface TableDef {
  table: TableSpec;
}

/**
 * Convert a TOML table definition to CREATE TABLE SQL.
 */
export function tomlToCreateSql(tomlStr: string): string {
  let def: TableDef;
  try {
    def = parse(tomlStr) as unknown as TableDef;
  } catch (e: unknown) {
    throw new MigrationError(
      `TOML parse error: ${e instanceof Error ? e.message : String(e)}`,
    );
  }

  if (!def.table || !def.table.name || !Array.isArray(def.table.columns)) {
    throw new MigrationError('TOML parse error: missing table definition');
  }

  const colDefs: string[] = [];
  const primaryKeys: string[] = [];

  for (const col of def.table.columns) {
    const parts: string[] = [`${col.name} ${col.type}`];

    const nullable = col.nullable ?? true;

    if (!nullable && !col.primary_key) {
      parts.push('NOT NULL');
    }

    if (col.unique) {
      parts.push('UNIQUE');
    }

    if (col.default !== undefined) {
      parts.push(`DEFAULT ${col.default}`);
    }

    if (col.references) {
      parts.push(`REFERENCES ${col.references}`);
    }

    if (col.primary_key) {
      primaryKeys.push(col.name);
    }

    colDefs.push(parts.join(' '));
  }

  if (primaryKeys.length > 0) {
    colDefs.push(`PRIMARY KEY (${primaryKeys.join(', ')})`);
  }

  return `CREATE TABLE ${def.table.name} (\n  ${colDefs.join(',\n  ')}\n);`;
}
